dali/operators/reader/loader/video/frames_decoder_gpu.cc

```diff
@@ -0,0 +1,190 @@
+    CUDA_CALL(cudaDeviceSynchronize());
+  }
+
+  CUDA_CALL(cuStreamSynchronize(0));
+  CUDA_CALL(cuvidUnmapVideoFrame(frames_decoder->nvdecode_state_->decoder, frame));
+
+  frames_decoder->decode_success_ = true;
+  return 1;
+}
+}  // namespace detail
+
+FramesDecoderGpu::FramesDecoderGpu(const std::string &filename) :
+  FramesDecoder(filename) {
+    nvdecode_state_ = std::make_unique<NvDecodeState>();
+
+    // Create nv decoder
+    CUVIDDECODECREATEINFO decoder_info = {};
+    memset(&decoder_info, 0, sizeof(CUVIDDECODECREATEINFO));
+
+    decoder_info.bitDepthMinus8 = 0;
+    decoder_info.ChromaFormat = cudaVideoChromaFormat_420;
+    decoder_info.CodecType = cudaVideoCodec_MPEG4;
+    decoder_info.ulHeight = Height();
+    decoder_info.ulWidth = Width();
+    decoder_info.ulMaxHeight = Height();
+    decoder_info.ulMaxWidth = Width();
+    decoder_info.ulTargetHeight = Height();
+    decoder_info.ulTargetWidth = Width();
+    decoder_info.ulNumDecodeSurfaces = 4;
+    decoder_info.ulNumOutputSurfaces = 1;
+
+    CUDA_CALL(cuvidCreateDecoder(&nvdecode_state_->decoder, &decoder_info));
+
+    // Create nv parser
+    CUVIDPARSERPARAMS parser_info;
+    memset(&parser_info, 0, sizeof(CUVIDPARSERPARAMS));
+    parser_info.CodecType = cudaVideoCodec_MPEG4;
+    parser_info.ulMaxNumDecodeSurfaces = 1;
+    parser_info.ulMaxDisplayDelay = 0;
+    parser_info.pUserData = this;
+    parser_info.pfnSequenceCallback = detail::process_video_sequence;
+    parser_info.pfnDecodePicture = detail::process_picture_decode;
+    parser_info.pfnDisplayPicture = nullptr;
+
+    CUDA_CALL(cuvidCreateVideoParser(&nvdecode_state_->parser, &parser_info));
+}
+
+void FramesDecoderGpu::SeekFrame(int frame_id) {
+  flush_ = true;
+  CUVIDSOURCEDATAPACKET *packet = &nvdecode_state_->packet;
+  memset(packet, 0, sizeof(CUVIDSOURCEDATAPACKET));
+  packet->payload = nullptr;
+  packet->payload_size = 0;
+  packet->flags = CUVID_PKT_ENDOFSTREAM;
+  CUDA_CALL(cuvidParseVideoData(nvdecode_state_->parser, packet));
+
+  last_frame_read_ = false;
+
+  FramesDecoder::SeekFrame(frame_id);
+}
+
+bool FramesDecoderGpu::ReadNextFrame(uint8_t *data, bool copy_to_output) {
+  decode_success_ = false;
+  current_copy_to_output_ = copy_to_output;
+  current_frame_output_ = data;
+
+  while (av_read_frame(av_state_->ctx_, av_state_->packet_) >= 0) {
+    if (av_state_->packet_->stream_index != av_state_->stream_id_) {
+      continue;
+    }
+
+    CUVIDSOURCEDATAPACKET *packet = &nvdecode_state_->packet;
+    memset(packet, 0, sizeof(CUVIDSOURCEDATAPACKET));
+    packet->payload = av_state_->packet_->data;
+    packet->payload_size = av_state_->packet_->size;
+    packet->flags = CUVID_PKT_TIMESTAMP;
+    packet->timestamp = av_state_->packet_->pts;
+
+    CUDA_CALL(cuvidParseVideoData(nvdecode_state_->parser, packet));
+
+    if (decode_success_) {
+      return true;
+    }
+  }
+
+  if (!last_frame_read_) {
+    CUVIDSOURCEDATAPACKET *packet = &nvdecode_state_->packet;
```

Comment:
I guess this could be a function as can be called in L147 as well.

Reply:
Done